### ISSUE TITLE
[doc, dir structure] Design spec / DV plan documentation with Hugo

### ISSUE THREAD

**Issue by sriyerg** [lowRISC/opentitan#751]
Seems like Hugo requires documentation pages that contain resources (such as images) to live in different directories. 

We initially placed both, the design specification as well as the DV plan under  `hw/ip/*/doc`. With Hugo, the DV plan got moved to a sub-directory called `dv_plan`. The landing page (URL) for `doc` now reflects the design specification. This makes it less obvious that the `dv_plan` even exists. 

It would be preferable to have the DV plan document to live in parallel to the design specification rather than be its leaf. 

I would like to propose the following possible solutions to this:

1. Do away with `hw/ip/*/doc` directory. Move the design specification under `hw/ip/*/rtl/spec` and DV plan under `hw/ip/*/dv/plan`. `data` directory will contain sources that are common to both, design and DV. 
  **Advantages:** All things design and all things DV are together in one place. URLs for both are super clear. 

2. Under `hw/ip/*/doc`, create sub-directory for `spec`
  **Advantages:** All documentation pieces live together in one place. Relatively less disruptive change. 
 
 Seeking thoughts and opinions on this. 

THanks
Sri

**Comment by gkelly** [comment]
@asb Doing a predictable munging like you're alluding to actually has the side benefit of us being able to filter the source better.  Right now we have to exclude a bunch of files from being included.  We can also ditch the pile of relative paths in the configs.

How does this sound for a proposal:
1. All `<NAME>.md` files are moved into directories of the same name as `index.md`.  Unless they're already named `index.md` or `_index.md` in which case we leave them.
2. We parse the markdown files and find images which they use as resources, and move them into their target directory.

Pros:
1. Links don't need to be updated, pages will have the same name (if they're not being linked to directly by their filename).  Their URLs stay the same.
2. We no longer have to worry about configuring the exclusion pattern.
3. Our `config.toml` gets a lot simpler

Cons:
1. Errors might not make as much sense, as they'll reference directories that don't exist anymore.

I've opened [a thread](https://discourse.gohugo.io/t/images-alongside-content/21649) on the hugo discourse server to see if anyone has any ideas in the meantime.

**Comment by asb** [comment]
I think there's no solution here that is particularly beautiful here. The choice is clearly between leaning more heavily into the way hugo prefers things, or to instead do a bit more preprocessing/munging so that hugo is more of an implementation detail to most contributors. Although I like hugo and think it's the right choice for us, I prefer the latter option for the in-tree docs. You're right that dealing with excluded files will also become much simpler. It's a shame that the hugo page bundle model is a bit of a mismatch for us. I'd be tempted to just conservatively copy all the images foo/ to foo/leaf1 and foo/leaf2 etc rather than bothering to try to identify which ones are actually referenced.

Something for a separate issue (or maybe we broaden the scope of this one), but if we're going to be doing this path changing anyway I'd be interested in:
* Using normal GitHub-compatible markdown links wherever possible, and doing a trivial rewrite to relref upon processing the markdown
* If title isn't specified in the frontmatter, then use the first `#` as title.

That kind of approach is more what I had in mind when advocating the use of hugo, but I appreciate tastes vary and there are some downsides to this. I think this sort of site/doc generation problem is just a bit inherently messy, messier than you would first think.

**Comment by eunchan** [comment]
On Thu, Nov 07, 2019 at 11:42:01PM -0800, Alex Bradbury wrote:
>I think there's no solution here that is particularly beautiful here. The choice is clearly between leaning more heavily into the way hugo prefers things, or to instead do a bit more preprocessing/munging so that hugo is more of an implementation detail to most contributors. Although I like hugo and think it's the right choice for us, I prefer the latter option for the in-tree docs. You're right that dealing with excluded files will also become much simpler. It's a shame that the hugo page bundle model is a bit of a mismatch for us. I'd be tempted to just conservatively copy all the images foo/ to foo/leaf1 and foo/leaf2 etc rather than bothering to try to identify which ones are actually referenced.
>
>Something for a separate issue (or maybe we broaden the scope of this one), but if we're going to be doing this path changing anyway I'd be interested in:
>* Using normal GitHub-compatible markdown links wherever possible, and doing a trivial rewrite to relref upon processing the markdown
>* If title isn't specified in the frontmatter, then use the first `#` as title.
>
>That kind of approach is more what I had in mind when advocating the use of hugo, but I appreciate tastes vary and there are some downsides to this. I think this sort of site/doc generation problem is just a bit inherently messy, messier than you would first think.
>
I agree with Alex's opinion. I am an advocate for the plain markdown, no fancy 
taggings, no trick to make it well rendered. I belive the markdown should be 
readiable in a raw text. That's why I prefer 80 characters per line in markdown 
(sorry for our markdown styleguide recommends one line per sentence..)

**Comment by msfschaffner** [comment]
@sriyerg can this issue be closed?
